jenkins_pipelines/environments/common/pipeline-build-validation.groovy

```diff
@@ -312,7 +320,10 @@ def clientTestingStages(capybara_timeout, default_timeout) {
     //Get minion list from terraform state list command
     def nodesHandler = getNodesHandler()
     def mu_sync_status = nodesHandler.MUSyncStatus
-
+    // Check persistent monitoring state
+    // Explanation : monitoring is using the same MU than one of the sles minion, we can't add twice the MUs
+    // We are using a persistent file to show monitoring MU has been added
+    def monitoring_bootstrap_output = sh(script: "test -e ${env.resultdir}/sumaform/monitoring && echo 'true' || echo 'false'", returnStdout: true).trim()
```

Comment:
Why don't handle this from inside the test framework?
We can only need to check if the MU has been added, if so, we skip the step.

That would very much simplify the complexity of the pipeline, moving this responsibility to the test framework.

Reply:
I'm not sure to understand how to move this to the test framework. 
Do you mean we you should always run the stage on the pipeline side and add steps int the testsuite that check if MU exist before adding it ? It would change drastically our testsuite on this side. For now we are failing if you try to add an existing MU. 
This behavior/ issue is just the consequence of reusing the same MU for monitoring and sles15sp4.
For me it's more a pipeline behavior and only the Jenkins worker can store if the monitoring call has been made before and no new deployment were executed again. And only the Jenkins worker can skip Rake call.

But I'm open to other idea if you have a better way to manage this. But you need to understand why we need it in the first place.

Reply:
```ruby
  desc 'Generate feature to add Maintenance Update Channels and repositories for all the clients'
  task :generate_add_maintenance_update_repository_features do
    clients = Rake::FileList['features/build_validation/init_clients/*.feature']
    clients.reject! { |client| client.include?('ssh') }
    clients.reject! { |client| client.include?('monitoring_server') }
```
we should already be doing that, since rake should skip over the monitoring server. Something needs to be fixed here if that's not the case :thinking: 
This is in our Rakefile, for reference.

Reply:
@maximenoel8 Maybe I missed something, but we already have two ways to prevent duplication, one as mentioned by Vanda is this reject method we have in some rake tasks.
The other is this code we have in the test framework:
```
When(/^I create the MU repositories for "([^"]*)"$/) do |client|
  repo_list = $custom_repositories[client]
  next if repo_list.nil?

  repo_list.each do |_repo_name, repo_url|
    unique_repo_name = generate_repository_name(repo_url)
    if repository_exist? unique_repo_name
      log "The MU repository #{unique_repo_name} was already created, we will reuse it."
    else
      content_type = deb_host?(client) ? 'deb' : 'yum'
      steps %(
        When I follow the left menu "Software > Manage > Repositories"
        And I follow "Create Repository"
        And I enter "#{unique_repo_name}" as "label"
        And I enter "#{repo_url.strip}" as "url"
        And I select "#{content_type}" from "contenttype"
        And I click on "Create Repository"
        Then I should see a "Repository created successfully" text or "The repository label '#{unique_repo_name}' is already in use" text
        And I should see "metadataSigned" as checked
      )
    end
  end
end
```

Note this part:
```
    if repository_exist? unique_repo_name
      log "The MU repository #{unique_repo_name} was already created, we will reuse it."
```